EIPS/eip-7916.md

```diff
@@ -38,30 +67,108 @@ ProgressiveByteList[3000, 4]
 
 ## Rationale
 
-`ByteList` definitions sometimes are defined with capacities that exceed their typical practical usage, to avoid limiting design space. For example, the [`Transaction`](https://github.com/ethereum/consensus-specs/blob/3c028dc73f5d93defc9bfd38c44784573a0bc70a/specs/bellatrix/beacon-chain.md#custom-types) type has a maximum capacity of 1 GB [`MAX_BYTES_PER_TRANSACTION`](https://github.com/ethereum/consensus-specs/blob/3c028dc73f5d93defc9bfd38c44784573a0bc70a/specs/bellatrix/beacon-chain.md#execution), despite actual transactions often being below 1 KB. This difference results in at least 20 extra hashes for common cases, to offer a design space that is not practically viable.
+### Why a Recursive Structure?
+
+- Efficiency: Small lists use fewer hashes (e.g., a 3-item list in a 16-element subtree wastes fewer hashes than a 1024-element `List[T, N]`).
+- Stability: Fixed subtree sizes ensure stable gindices, avoiding the need for dynamic depth adjustments or multiple queries.
+- Scalability: Recursive subtrees allow arbitrary growth without a hardcoded limit, constrained only by practical limits (e.g., network payload limit, validation rules).
+
+### Why Not Dynamic Depth?
+
+Dynamic-depth Merkleization destabilizes gindices:
+
+- Requires two-step queries (length then gindex), increasing latency and reorg risks.
+- Complicates proofs with semantic lookups.
+
+Mixing in successor subtrees ensures predictable gindices and proof sizes.
+
+### Why Not Fixed-Capacity Lists?
+
+`List[T, N]`:
+
+- Imposes arbitrary limits, hindering scalability.
+- Breaks stability when redefined.
+- Wastes hashes with padding (e.g., 1024-element capacity for a 1-item list). (only log(N) wasted hashes)
+
+`ProgressiveList[T]` offers a scalable, efficient alternative.
+
+### Why Are Base Size and Scaling Factors Not Exposed Parameters?
 
-With this EIP, common cases can be hashed more efficiently while not restricting the design space, and while also retaining stable generalized Merkle indices for each individual chunk of the `ByteList`.
+- Simplicity: Fixed values (base size 1, scaling factor 4) provide a sensible default that balances efficiency and usability, aligning with SSZ’s goal of simplicity.
+- Future Extensibility: If specific use cases demand different values, a future EIP could introduce parameterization. For now, fixed values reduce adoption barriers and align with the principle of "good enough" defaults.
 
 ## Backwards Compatibility
 
-The new SSZ type does not interfere with any existing types.
+`ProgressiveList[T]` is a new SSZ type, coexisting with `List[T, N]` and other types without conflict. Its `List`-equivalent serialization ensures compatibility with existing serializers.
 
 ## Security Considerations
 
-None
+- Resource limits: The `uint32` length-prefix caps serialization of variable-length subsections at 4GB, but practical limits (e.g., 10MB libp2p messages) apply. Implementations SHOULD enforce context-specific bounds.
```

Comment:
Actually, why define it as `uint32`? I mean it has little practical importance, but the length is a full chunk and encoded in little endian anyways, so it could just as well be `uint256`. Practical limits should still be enforced and they will be typically lower than 4G elements but since the type definition says it's a limitless list, I feel like it would just look better.

Reply:
The `uint32` is just how offsets get encoded of variable-length sections within a complex type. Have updated the wording to better reflect that.